### ISSUE TITLE
Windows Unicode

### ISSUE THREAD

**Issue by degski** [carlomilanesi/cpp-mmf#1]
Does not compile with Unicode (the default, and should be used imo): cannot convert const wchar_t * to const char *. Compiling with option Multibyte code does make this compile, but since we are dealing with the filesystem, this hardly seems to be a good idea.

**Comment by carlomilanesi** [comment]
I presume the error happens when you pass a `const wchar_t *` to the constructor of `read_only_mmf`, expecting a `const char *`. I am using Linux now. For Windows, that argument is used as the the first argument of a call to `CreateFile`, whose expected type is `LPCTSTR`, whose value is `const char *`. With Linux, it accepts Unicode as UTF-8. Could you tell me how to open a file in Windows using a `const wchar_t *` as pathname?

**Comment by degski** [comment]
Yes, it's the `LPCTSTR` that causes the problem. "Could you tell me how to open a file in Windows using a const wchar_t * as pathname?" A wchar_t has size 2, that's why it cannot work any other way. You'll have to pass it a `std::wstring`, a `wchar_t` (duh), or a wide string-literal `L"foo.ext"`. In essence though, you'd have to template to whole thing on a CharType. The windows CRT has 2 functions for everything (if relevant of course), often a macro that substitutes the correct one. In C++17 you could use `constexpr if`. It's [the ~~bloody~~ wide/narrow divide] quite a nightmare in general, and usually I ignore things, but in the case of mmf, that seems unwise, as the filesystem itself (NTFS), is fully unicode (it's the windows layer on top that creates the mess and usually for historic reason). Even if you would use `<filesystem>` in C++17, you'd still have this problem. Untill the language gives full support for unicode, it's gonna stay that way.

**Comment by myd7349** [comment]
Using `CreateFileA` instead of `CreateFile` will fix this issue.

**Comment by myd7349** [comment]
@carlomilanesi `LPCTSTR` is not always `const char *`. When `UNICODE` is defined, `LPCTSTR` will be `const wchar_t *`.
In fact, `CreateFile` is just a macro defined in `WinBase.h`(see #2 ).
And if you want to open a file with a file name of type `const wchar_t *`, `CreateFileW` is what you are looking for.

**Comment by myd7349** [comment]
@degski If you have a file name of `LPCTSTR`, you may convert it into a `const char *` by using `T2A` macro in case that you are working on a `ATL` or `MFC` project. Otherwise, you may use `WideCharToMultiByte` to do this convertion manually.

[T2A](https://msdn.microsoft.com/en-us/library/87zae4a3.aspx)
[WideCharToMultiByte](https://docs.microsoft.com/en-us/windows/desktop/api/stringapiset/nf-stringapiset-widechartomultibyte)

**Comment by degski** [comment]
@myd7349 

Originally I wrote: "Compiling with option Multibyte code does make this compile", it does `WideCharToMultiByte` for the whole project, yes it works. This was not what I was talking about, the best would be to have  overloads for both types, so the whole thing would be unicode, that was my point.

**Comment by myd7349** [comment]
@degski Got it. You mean we need an `open(const wchar_t *)`. I didn't read it carefully, sorry for that. :-)

**Comment by degski** [comment]
@myd7349 

Yes, that's what I mean. Natively it's unicode all the way (MS recommends to compile with UNICODE!), so the conversion, really does a conversion, and back, so you can imagine that that is not gonna be optimal.
